Fix Apple Silicon shuffle thunks

sdmaclea opened the pull request:
Fixes #47294

sdmaclea committed:
Fix Apple Silicon shuffle thunks

Fixes 47294

sdmaclea commented:
I ran the three failing tests locally and they all passed.

AaronRobinsonMSFT commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -189,16 +189,57 @@ class ShuffleIterator
+            // Only Apple Silicon ABI currently supports unaligned stack argument shuffling
+            _ASSERTE(byteIndex == index * TARGET_POINTER_SIZE);
+            return index;
+#else
```
Is there a doc reference for this logic?

sdmaclea replied:
This is just a loop to emit stack load/store `ShuffleEntry`s.  Arm64 logic can write 1,2,4,or 8 bytes in a single instruction.

I'll add a comment.

AaronRobinsonMSFT commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -189,16 +189,57 @@ class ShuffleIterator
         if (m_currentByteStackIndex < m_argLocDesc->m_byteStackSize)
         {
             const unsigned byteIndex = m_argLocDesc->m_byteStackIndex + m_currentByteStackIndex;
```
Include this in the `#if` since it is not used anywhere else in this branch.

sdmaclea replied:
It is used in the else branch.

AaronRobinsonMSFT commented:
/cc @jkoritzinsky @elinor-fung

janvorli commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -189,16 +189,57 @@ class ShuffleIterator
             m_currentByteStackIndex += TARGET_POINTER_SIZE;
 
             // Delegates cannot handle overly large argument stacks due to shuffle entry encoding limitations.
+            index = byteIndex / TARGET_POINTER_SIZE;
```
A nit - can you please move it to the original location few lines above as this change has no functional effect?

janvorli commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -189,16 +189,57 @@ class ShuffleIterator
                 COMPlusThrow(kNotSupportedException);
             }
 
-            return -(int)byteIndex;
```
Can you please fix the function description comment? This change makes it invalid.

janvorli commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -189,16 +189,57 @@ class ShuffleIterator
+            _ASSERTE(byteIndex == index * TARGET_POINTER_SIZE);
+            return index;
+#else
+            int bytesRemaining = m_argLocDesc->m_byteStackSize - m_currentByteStackIndex;
```
I think this mechanism needs a comment describing what it does. It is not clear to me, as arguments can only be of a  power of two size, but this seems to handle any arbitrary size and splits it to a sequence of TARGET_POINTER_SIZE sized blocks followed by blocks of 4, 2 and 1 bytes sizes (only the ones that are needed to form the target size).

I also wonder if it honors the following from the Apple doc:
> Function arguments may consume slots on the stack that are not multiples of 8 bytes. If the total number of bytes for stack-based arguments is not a multiple of 8 bytes, insert padding on the stack to maintain the 8-byte alignment requirements. 

Imagine a case when all arguments are passed in registers in the source signature, using all the available general purpose registers. The last argument is of a single byte size and we insert a "this" reference as the first argument in the destination signature. The last argument then goes to a stack slot and the stack needs to be aligned to 8.

sdmaclea replied:
Last I knew the arm64 stack needed to always be 16-byte aligned (to support ld/st pair & simd)

The existing code does not support an argument moving from a register onto the stack.  Only the opposite case is supported (stack -> register). See
https://github.com/dotnet/runtime/blob/c95aa3f48aa591dca870ee00c31c3d8bdc740b5a/src/coreclr/vm/arm64/stubs.cpp#L1720

So the stack only shrinks.... I don't think this code needs to do anything for padding or stack alignment.

I suspect the 3/5/6/7 cases are not really necessary.  It seemed reasonable to just implement them, but I could `_ASSERTE` instead.  My intent is to copy the argument stack size as reported by the `ArgLocDesc`

sandreenko replied:
> I suspect the 3/5/6/7 cases are not really necessary. It seemed reasonable to just implement them, but I could _ASSERTE instead. My intent is to copy the argument stack size as reported by the ArgLocDesc

I would prefer an assert.

sdmaclea replied:
I am looking at Vector3Interop it is hitting one of the new asserts.

I will rework this when I fix that failure.

sdmaclea replied:
For the failing Vector3Interop case, the Vector3 argument is

```console
print *m_argLocDesc
(ArgLocDesc) $2 = {
  m_idxFloatReg = -1
  m_cFloatReg = 0
  m_idxGenReg = -1
  m_cGenReg = 0
  m_byteStackIndex = 28
  m_byteStackSize = 12
  m_hfaFieldSize = 0
```

I am surprised it is not 8-byte aligned...

sandreenko replied:
is it hfa<float>?

sdmaclea replied:
Based on the alignment, it seems to be treated as HFA.  But I am surprised that `m_hfaFieldSize` is not `4` since it seems to be HFA.

I pushed a hack which inferred the alignment from the m_byteStackSize.  It feels wrong, but I don't really have a better idea.  I suspect it is sufficient to work for the current test coverage.

I wonder if `m_hfaFieldSize` should change to `m_stackAlignment`...

sandreenko replied:
so we expect hfaFieldSize to be set on stack passed arguments? It makes sense, maybe this line https://github.com/dotnet/runtime/blob/ad6a69735d61a1bc22da9aff15ebc5df94f63beb/src/coreclr/vm/callingconvention.h#L674-L683
prevents `setHFAFieldSize` from being set?

sdmaclea replied:
I was thinking it needed to be set here 

https://github.com/dotnet/runtime/blob/ad6a69735d61a1bc22da9aff15ebc5df94f63beb/src/coreclr/vm/callingconvention.h#L715-L718

sandreenko replied:
yes, I think it will work.

sdmaclea replied:
OK I revised using `m_hfaFieldSize`.  I think it is a correct fix.  I am resolving this .  PTAL again.

sdmaclea committed:
Feedback

sdmaclea committed:
Fix case

sdmaclea committed:
Fix comparison warning

sdmaclea committed:
Fix Vector3Interop test

Vector3 is a 12 byte structure with 4 byte alignment. We cannot assume the structure is 8 byte aligned on the stack so
we must move in 4 byte pieces.

Generalize this assumption.

sdmaclea committed:
Set and use m_hfaFieldSize for stack arguments

m_hfaFieldSize is needed to calculate correct shuffle size.

sdmaclea committed:
Fix typo

sdmaclea commented:
PTAL.

janvorli reviewed (approved):
LGTM, thank you!

sandreenko commented on src/coreclr/vm/comdelegate.cpp:
```diff
@@ -147,8 +147,11 @@ class ShuffleIterator
-    // For register arguments it returns regNum | ShuffleEntry::REGMASK | ShuffleEntry::FPREGMASK.
-    // For stack arguments it returns stack offset in bytes with negative sign.
+    // It returns an offset encoded properly for a ShuffleEntry offset.
+    // - For floating resgister arguments it returns regNum | ShuffleEntry::REGMASK | ShuffleEntry::FPREGMASK.
```
```suggestion
    // - For floating register arguments it returns regNum | ShuffleEntry::REGMASK | ShuffleEntry::FPREGMASK.
```

sandreenko reviewed (approved):
LGTM

sdmaclea committed:
Fix typo
